course_info/icommons.py

```diff
@@ -1,82 +1,116 @@
+                response = self.make_request('GET', relative_url,
+                                             headers=self.headers)
+                results = response.data.get('results', [])
+                if len(results):
+                    course_instance_id = self._get_course_instance_id(results)
+                    if course_instance_id:
+                        course_info = self.get_course_info(course_instance_id)
+                log_msg = "Caching course info for canvas_course_id {}: {}"
+                log.debug(log_msg.format(canvas_course_id, json.dumps(course_info)))
+                cache.set(cache_key, course_info)
+            except Exception as e:
+                log.exception(e.message)
         return course_info
 
     def get_school_info(self, school_id):
-        '''
+        """
             retrieves - from the iCommons API - the information
             about the school passed in (by its school id),
             or logs an error and returns an empty dict if there is no such school
-        '''
-        school_info = {}
-        try:
-            relative_url = '/schools/'+ school_id +'/?format=json'
-            response = self.make_request('GET', relative_url, headers=self.headers)
-            school_info = response.data.copy()
-        except drest.exc.dRestRequestError as e:
-            log.error(e.message)
-        except Exception as e:
-            log.error(e.message)
+        """
+        cache_key = CACHE_KEY_SCHOOL_BY_SCHOOL_ID.format(school_id)
+        school_info = cache.get(cache_key)
+        if school_info is None:
+            school_info = {}
+            try:
+                relative_url = '/schools/%s/?format=json' % school_id
+                response = self.make_request('GET', relative_url,
+                                             headers=self.headers)
+                school_info = response.data
+                log_msg = "Caching school info for school_id {}: {}"
+                log.debug(log_msg.format(school_id, json.dumps(school_info)))
+                cache.set(cache_key, school_info)
+            except Exception as e:
+                log.error(e.message)
         return school_info
         # This function could be made to return only the school name or an error, as that's the only thing we need it
         # for right now (in views.py, to insert the school name before the display number).
         # But just in case we want to use it for more than that in the future, and so that it mirrors 'get_course_info,'
         # we're going to let it return the whole data set.
+
+    def _get_course_instance_id(self, course_instances):
+        if not course_instances:
+            return None
+        if len(course_instances) == 1:
+            # Only one course found with the requested canvas course id;
+            # we'll use its course instance ID for the course info lookup
+            return course_instances[0]['course_instance_id']
```

Comment:
It might be paranoia, but it's theoretically possible that a secondary course might have the correct canvas course id set on it, when the primary doesn't.  This basically just wants to mirror the logic from `CourseInstanceManager.get_primary_course_by_canvas_course_id()`, right?